readme.md
## PyTMX
##### For Python 2.7 and 3.3+

This is the most up-to-date version of PyTMX available and works with
Python 2.7 and 3.3+.

If you have any problems or suggestions, please open an issue.
I am also often lurking #pygame on freenode.  Feel free to contact me.

Requires the six module.

*Released under the LGPL v3*

### See the "apps" folder for example use and cut/paste code.


News
===============================================================================

__10/27/16__ - Experimental pysdl2_cffi binding support  
__11/13/15__ - Animations are now loaded  
__07/08/15__ - Documentation overhaul  
__04/18/15__ - Document support for pysdl2 and pyglet  
__09/14/14__ - Merge python3 branch.  Now 100% compatible with 2.7 and 3.3+  
__07/26/14__ - New python3/2 release.  Check it out in the python3 branch.  
__05/29/14__ - Added support for rotated objects and floating point  
__04/04/14__ - New Six Branch created  
__02/28/14__ - Image layer support, object points changed, new test.py!  
__02/24/14__ - New Python 3 Support: see python3 branch  
__02/06/14__ - Python 3 support coming soon  


Introduction
===============================================================================

PyTMX is a map loader for python/pygame designed for games.  It provides smart
tile loading with a fast and efficient storage base.  Not only does it
correctly handle most Tiled object types, it also will load metadata for
them so you can modify your maps and objects in Tiled instead of modifying
your source code.

New support for pysdl2 and pyglet!  Check it out!

Because PyTMX was built with games in mind, it differs slightly from Tiled in
a few minor aspects:

- Layers not aligned to the grid are not supported.
- Some object metadata attribute names are not supported (see "Reserved Names")

PyTMX strives to balance performance and flexibility.  Feel free to use the
classes provided in pytmx.py as superclasses for your own maps, or simply
load the data with PyTMX and copy the data into your own classes with the api.

There is no save feature.  Once the map is loaded, it will be up to
you to provide a way to save changes to the map.  I've used the pickle module
with good results.

I need to clarify a few things:
- pytmx is not a rendering engine
- pytmx is not the Tiled Map Editor


Documentation
===============================================================================

This readme does include some detailed documentation, but the full API reference
and documentation can be found at the site below.  For examples of real use,
check out the apps folder in this repo.  The 'test' apps demonstrate how to
load maps, get layer, tile, and object data, as well as some rendering.

http://pytmx.readthedocs.org/


# Table of Contents
1. [Installation](#installation)
2. [Basic Use](#basic-use)
3. [Getting Properties](#object-properties)
4. [Working with Maps](#working-with-maps)
5. [Loading from XML](#loading-from-xml)
6. [Custom Image Loading](#custom-image-loading)
7. [Working with Tile Layers](#working-with-tile-layers)
8. [Getting Tile Animations](#getting-tile-animations)
9. [Working with Objects](#working-with-objects)
10. [Understanding Properties](#understanding-properties)


Getting Help
===============================================================================

For bugs or feature requests, please use the issues feature of github.  For
all other general questions, join me on IRC at freennode.net #pygame.


Design Goals and Features
===============================================================================

* API with many handy functions
* Memory efficient and performant
* Loads data, "properties" metadata, and images from Tiled's TMX format
* Supports base64, csv, gzip, zlib and uncompressed XML formats
* Properties for all native Tiled object types
* Point data for polygon and polyline objects
* Automatic flipping and rotation of tiles
* Built-in image loading with pygame, pysdl2, and pyglet
* Loads animation information


Why use PyTMX?
===============================================================================

### PyTMX is efficient:
* Only the tiles used on a map are loaded into memory
* Map information is stored as integers, not python objects (32+kb)
* Extensive use of generators and iterators make it easy on memory
* Code is designed for compact size and readability

### PyTMX is flexible:
* Supports all major Tiled features and object types
* PyTMX data classes can be extended
* Does not force you to render data in any particular way
* Includes many checks to give useful debugging information
* Supports pygame, pyglet, and pysdl2 image loading

### PyTMX is supported:
* GitHub hosting allows for community participation
* I have kept PyTMX current with new versions of Tiled since v.7

### PyTMX is usable:
* Liberal LGPL license means you can use PyTMX for your project


Installation
===============================================================================

Install from pip

    pip install pytmx


You can also manually install it

    python setup.py install


Basic use:
===============================================================================

#### Just data, no images:

```python
import pytmx
tiled_map = pytmx.TiledMap('map.tmx')
```

#### Load with pygame images:

```python
from pytmx.util_pygame import load_pygame
tiled_map = load_pygame('map.tmx')
```

#### Load with pysdl2 images (experimental):

```python
from pytmx.util_pysdl2 import load_pysdl2
tiled_map = load_pysdl2('map.tmx')
```

#### Load with pyglet images (experimental):

```python
from pytmx.util_pyglet import load_pyglet
tiled_map = load_pyglet('map.tmx')
```

#### Load from XML string:

```python
import pytmx
tiled_map = pytmx.TiledMap.from_xml_string(some_string_here)
```

#### Iterate through layers and groups:

```python
for layer in tiled_map.layers:
   ...
```

#### Iterate through tile images in a tile layer:

```python
for x, y, image in layer.tiles():
   ...
```

#### Iterate through Tiled objects in an object group:

```python
for obj in layer:
   ...
```

#### Get properties of various object types:

```python

# properties is a dict 
TiledMap.properties
TiledTileLayer.properties['name']
TiledObject.properties['type']

# tile ('GID') properties are accessed through the TiledMap:
properties = TiledMap.get_tile_properties(x, y, layer)
```

#### Get bounding box of an object:

```python
bbox = obj.x, obj.y, obj.width, obj.height
```

#### Get the points/vertex to draw a polygon/polyline object:

```python
points = obj.points
# if obj.closed == True, then obj is a polygon
```

Working with Maps
===============================================================================

TiledMap objects are returned from the loader.  They contain layers, objects,
and a bunch of useful functions for getting information about the map.  In
general, all of the pytmx types are not meant to be modified after being
returned from the loader.  While there is a potentional for modifing them,
its not a supported function, and may change any time.  Please consider them
read-only.

Here is a list of attributes for use.  (ie: TiledMap.layers):

- layers: all layers in order
- tile_properties: dictionary of tile properties {GID: {props...}, ...}
- layernames: dictionary of layers with names: {name: layer, ...}
- images: list of all images in use, indexed by GID.  Index 0 is always None.
- version
- orientation
- width: width of map in tiles, not pixels
- height: height of map in tiles, not pixels
- tileheight: height of tile in pixels.  may differ between layers.
- tilewidth: width of tile in pixels.  may differ between layers.
- background_color: map background color specified in Tiled
- properties: all user created properties about the map


#### Optional loading flags

All loaders support the following flags:
- load_all_tiles: if True, all tiles will be loaded, even if unused
- invert_y: used for OpenGL graphics libs.  Screen origin is at lower-left
- allow_duplicate_names: Force load maps with ambiguous data (see 'reserved names')

```python
from pytmx.util_pygame import load_pygame
tiled_map = load_pygame(path_to_tmx_file, invert_y=True)
```

#### Loading from XML

Most pytmx objects support loading from XML strings.  For some objects, they require
references to other objects (like a layer has references to a tileset) and won't load
directly from XML.  They can only be loaded if the entire map is loaded first.  If you
want to store XML in a database or something, you can load the entire map with an XML string:

```python
import pytmx
tiled_map = pytmx.TiledMap.from_xml_string(some_string_here)
```

#### Custom Image Loading

The pytmx.TiledMap object constructor accepts an optional keyword "image_loader".  The argument should be a function that accepts filename, colorkey (false, or a color) and pixelalpha (boolean) arguments.  The function should return another function that will accept a rect-like object and any flags that the image loader might need to know, specific to the graphics library.  Since that concept might be difficult to understand, I'll illustrate with some code.  Use the following template code to load images from another graphics library.

 ```python
import pytmx

def other_library_loader(filename, colorkey, **kwargs):

    # filename is a file to load an image from
    # here you should load the image in whatever lib you want

    def extract_image(rect, flags):
    
        # rect is a (x, y, width, height) area where a particular tile is located
        # flags is a named tuple that indicates how tile is flipped or rotated
    
        # use the rect to specify a region of the image file loaded in the function
        # that encloses this one.
        
        # return an object to represent the tile
        
        # what is returned here will populate TiledMap.images, be returned by
        # TiledObject.Image and included in TiledTileLayer.tiles()

    return extract_image

level_map_and_images = pytmx.TiledMap("leveldata.tmx", image_loader=other_library_loader)
```

#### Accessing layers

Layers are accessed through the TiledMap class and there are a few ways to get references to them:

```python
# get a layer by name
layer_or_group = tiled_map.get_layer_by_name("base layer")

# TiledMap.layers is a list of layers and groups
layer = tiled_map.layers[layer_index_number]

# easily get references to just the visible tile layers
for layer in tiled_map.visible_tile_layers:
    ...

# just get references to visible object groups
for group in tile_map.visible_object_groups:
    ...
```


Working with Tile Layers
===============================================================================

Pytmx loads tile layers and their data:

- name
- opacity
- visible: indicates if user has hidden the layer
- data: 2d array of all tile gids (normally not needed to use!)
- properties

#### Tile Images

Single tile images are accessible from TiledMap, TiledTileLayer, and TiledObject objects.
If you requre all images in a layer, there are more effecient ways described below.

```python
# get image from the TiledMap using x, y, and layer numbers
pygame_surface = tile_map.get_tile_image(x, y, layer)

# get tile image from an object with a image/GID assigned to it
image = obj.image

# get image using gid (not needed for normal use!)
gid = layer.data[y][x]
image = tiled_map.images[gid]
```

#### Least effort involved getting all tile images

```python
for x, y, image in layer.tiles():
    ...
```

#### Getting tile animations

Tiled supports animated tiles, and pytmx has the ability to load them.
Animations are stored in the properties for the tile.  Animations from
pytmx are a list of AnimationFrame namedtuples.  Please see the example below.

```python
# just iterate over animated tiles and demo them

# tmx_map is a TiledMap object
# tile_properties is a dictionary of all tile properties

# iterate over the tile properties
for gid, props in tmx_map.tile_properties.items():

   # iterate over the frames of the animation
   # if there is no animation, this list will be empty
   for animation_frame in props['frames']:
   
       # do something with the gid and duration of the frame
       # this may change in the future, as it is a little awkward now
       image = tmx_map.get_tile_image_by_gid(gid)
       duration = animation_frame.duration
       ...

```

#### If you really want to work with layer data directly...

This information is provided for the curious, but for most people is not
required for normal use.

Layer tiles are stored as a 'list of lists', or '2d array'.  Each element of
layer data is a number which refers to a specific image in the map.  These
numbers are called GID.  Do not make references to these numbers, as they
will change if the map changes. 

Images for the GID can be accessed with the TiledMap.images list.

With pygame, images will be plain pygame surfaces.  These surfaces will be
checked for colorkey or per-pixel alpha automatically using information from
the TMX file and from checking each image for transparent pixels.  You
do not need, and should not convert the tiles because it is already done. 

```python
layer = tiled_map.layers[0]
for x, y, gid in layer:
    ...

# get image using gid (not needed for normal use!)
# row index = 'y'
# column index = 'x'
image_gid = layer[row_index][column_index]
image = tiled_map.images[image_gid]

# change GID of a position
layer[y][x] = new_gid
```

Working with Objects
===============================================================================

Tiled "objects" are things that are created in object layers, and include
polygons, polylings, boxes, ellispes, and tile objects.  Pytmx loads all objects
and their data:

- name
- type
- x
- y
- width
- height
- rotation
- gid (if it has an image)
- visible
- image
- properties

#### Basics
Attributes x, y, width, and height all represent the bounding box of the object,
even polygons and polylines.

#### Image Objects
If using a loader, then TiledObject.image will be a reference to the image used.

#### Tile Objects
Tile Objects are objects that reference a tile in a tileset.  These are loaded and
the image will be available to use.

#### Polygon/Polyline Objects
These objects have special attributes: 'closed' and 'points'.  Each point is (x, y) tuple.
If the object is a polygon, then TiledObject.closed will be True.  Points are not
rotated if the rotation property is used.

#### Accessing objects

Objects can be accessed through the TiledMap or through a group.  Object groups can be
used just like a python list, and support indexing, slicing, etc.

```python
# search for an object with a specific name
my_object = tiled_map.get_object_by_name("baddy001")  # will not return duplicates

# get a group by name
group = tiled_map.get_layer_by_name("traps")

# copy a group
traps = group[:]

# iterate through objects in a group:
for obj in group:
    ...
```

Understanding Properties
===============================================================================

Properties are a powerful feature of Tiled that allows the level designer to
assign key/value data to individual maps, tilesets, tiles, and objects.  Pytmx
includes full support for reading this data so you can set parameters for stuff
in Tiled, instead of maintaining external data files, or even values in source.

Properties are created by the user in tiled.  There is also another set of data
that is part of each object, accessed by normal object attributes.  This other
data is not set directly by the user, but is instead set by tiled.  Typical
data that is object attributes are: 'name', 'x', 'opacity', or 'id'.

If the user sets data for an object in Tiled, it becomes part of 'properties'.
'Properties' is just a normal python dictionary.

```python
# get data normally set by Tiled
obj.name
obj.x
obj.opacity

# get data set by the user in Tiled
obj.properties['hit points']
obj.properties['goes to 11']
```

Individual tile properties are accessed through the the parent map object:

```
tiled_map = TiledMap('level1.tmx')
props = tiled_map.get_tile_properties(x, y, layer)
props = tiled_map.get_tile_properties_by_gid(tile_gid)
```


Scrolling Maps for Pygame
===============================================================================

I have another repo with a working demo of a proper scrolling map using Tiled
maps and pygame.  Please feel free to test drive it.  It isn't limited to Tiled
maps, you can use any data structure you want, as long as pygame is used.

https://github.com/bitcraft/pyscroll


Reserved Names
================================================================================

Tiled supports user created metadata called "properties" for all the built-in
objects, like the map, tileset, objects, etc.  Due to how the Tiled XML data is
stored, there are situations where Tiled internal metadata might have the same
name as user-created properties.

Pytmx will raise a ValueError if it detects any conflicts.  This check is
performed in order to prevent any situations where a level change might be made

changelog
Map loader for TMX Files
Leif Theden, "bitcraft", 2012-2016

v3.20.x - for python 2.7 and 3.3+
Tested with Tiled 0.11.0.

Released under the LGPL v3

===============================================================================

New in 3.20:
      core: option to disable duplicate property checks: allow_duplicate_names
      core: keyword arguments correctly passed with util_* loaders
      core: new convenience methods to simplify drawing
      core: more abstract loading of images
      core: tile images can be reloaded after tmx data is loaded
      core: new keyword: "invert_y", to support the pyglet loader
    pysdl2: support image loading with pysdl2 (experimental)
    pyglet: support image loading with pyglet (experimental)

New in 3.19:
      core: new iterator for tiledmap
      core: objects/shapes load the rotation value
      core: polygon/polyline and all other shapes return coordinates as float
      core: pytmx respects tiles that specify external image: stored in metadata
      core: tileoffsets are loaded stored in tileset.offset: tuple: (x, y)
      core: logging module used everywhere
    pygame: tilesets can be loaded even if they don't specify an image
    pygame: loader has new logger...more descriptive error messages
    pygame: loading of tiles that specify an external image is supported
    pygame: new optional arguments for load_pygame():
            load_all: bool (False is default), load al tiles, even unused ones
            optional_gids: list/tuple, also load the gids in this list
                           'gid' refers to the gid found in tiled

New in 3.18:
    pygame: removed option for force a colorkey for a tileset
    pygame: pixelalpha is now enabled by default
     pytmx: Maps can now be loaded from pytmx.TiledMap.fromstring(xml_string)
     pytmx: pygame is no longer a required dependency (to be tested!)
      core: Sphinx documentation created

New in 3.17:
    loader: removed legacy load_tmx function: just call TiledMap() instead
    loader: added test to correct tilesheets that include non-tile graphics
     pytmx: polygon objects now return absolute coordinates in points
     pytmx: tiled properties are now available through dictionary "properties"
      core: tested with the mana world maps...it works!
      demo: simplified the demo/test for easier readability
      test: maps now render and are scaled inside the window to show entire map


New in 3.16:
    ***    jumped to version 3.x to reflect new python 3.3 compatibility    ***

       all: python 3 support
      pep8: changed method/function names to lowercase with underscore spacing
      pep8: modified various style infractions
      core: simplified file structure
      core: added __all__ to some modules for less clutter
      demo: added ability to resize preview window
      test: mouse clicks now advance the test
      test: added ability to resize preview window
      test: tile objects are drawn (previously supported, but not shown in test)
     utils: renamed buildDistributionRects to build_rects
     pytmx: bumped up gid limit from 255 to 65535 (16-bit)
     pytmx: removed get_objects(), replaces with objects property
     pytmx: removed get_draw_order(), replaced with visible_layers property
    loader: added ImageLayer support
    loader: minor documentation fixes
    loader: small optimizations
    loader: possible [ultra minor] optimization: using iterator on etree to load

New in .15:
    loader: new getTileLayerByName(name) method
    loader: python 2.6 support
    loader: fixed issue where objects with tile gid did not load properties
    loader: polygon and polyline objects
    loader: new lookup methods use iterators
    loader: loading function moved into classes
    loader: data/images can be reloaded on the fly
    loader: uses etree for faster xml parsing

New in .14:
    loader: Fixed gid lookup for "buildDistributionRects"
    loader: Added useful output to a few classes "__repr__"
    loader: Fixed a gid mapping issue that broke rotated tiles
    pygame: fixed colorkey handling
    pygame: correctly handles margins and spacing between tiles in tilesets
    pygame: b/c of changes, now correctly renders tiled's example maps
    added scrolling demo

New in .13:
    loader: Renamed "get_tile_image" to "getTileImage"
    loader: Removed duplicates returned from getTilePropertiesByLayer
    loader: Modified confusing messages for GID errors
    loader: Fixed bug where transformed tile properties are not available
    loader: No longer loads metadata for tiles that are not used
    loader: Reduced tile cache to 256 unique tiles
    loader: Removed 'visible' from list of reserved words
    loader: Added 'buildDistributionRects' and maputils module
    loader: Added some misc. functions for retrieving properties
    pygame: Smarter tile management made tile loading cache useless; removed it
    pygame: pygame.RLEACCEL flag added when appropriate

New in .12:
    loader: Fixed bug where tile properties could contain reserved words
    loader: Reduced size of image index by only allocating space for used tiles

New in .11:
    loader: Added support for tileset properties
    loader: Now checks for property names that are reserved for internal use
    loader: Added support for rotated tiles
    pygame: Only the tiles that are used in the map will be loaded into memory
    pygame: Added support for rotated tiles
    pygame: Added option to force a bitsize (depth) for surfaces
    pygame: Added option to convert alpha transparency to colorkey transparency
    pygame: Tilesets no longer load with per-pixel alphas by default
    pygame: Colorkey transparency should be correctly handled now

===============================================================================


pytmx/util_xml.py
import struct
import array
import logging
import os.path
import six.moves
from six.moves import zip
from six.moves import map
from itertools import product
from collections import OrderedDict
from collections import namedtuple
from collections import defaultdict
from xml.etree import ElementTree

from pytmx import *

__all__ = ['parse_map',
           'parse_tileset',
           'parse_layer',
           'parse_objectgroup',
           'parse_object',
           'parse_imagelayer',
           'parse_properties',
           'from_xml']

logger = logging.getLogger(__name__)
ch = logging.StreamHandler()
ch.setLevel(logging.INFO)
logger.addHandler(ch)
logger.setLevel(logging.INFO)


def convert_to_bool(text):
    # properly convert strings to a bool
    try:
        return bool(int(text))
    except:
        pass

    text = str(text).lower()
    if text == "true":
        return True
    if text == "yes":
        return True
    if text == "false":
        return False
    if text == "no":
        return False

    raise ValueError

# used to change the unicode string returned from xml to proper python types.
types = defaultdict(lambda: str)
types.update({
    "version": float,
    "orientation": str,
    "width": int,
    "height": int,
    "tilewidth": int,
    "tileheight": int,
    "firstgid": int,
    "source": str,
    "name": str,
    "spacing": int,
    "margin": int,
    "trans": str,
    "id": int,
    "opacity": float,
    "visible": convert_to_bool,
    "encoding": str,
    "compression": str,
    "gid": int,
    "type": str,
    "x": float,
    "y": float,
    "value": str,
    "rotation": float,
})

# internal flags
TRANS_FLIPX = 1
TRANS_FLIPY = 2
TRANS_ROT = 4

# Tiled gid flags
GID_TRANS_FLIPX = 1 << 31
GID_TRANS_FLIPY = 1 << 30
GID_TRANS_ROT = 1 << 29

flag_names = (
    'flipped_horizontally',
    'flipped_vertically',
    'flipped_diagonally',)

TileFlags = namedtuple('TileFlags', flag_names)


def from_xml(filename, tiled_map):
    etree = ElementTree.parse(filename).getroot()
    parse_map(etree, tiled_map)


def handle_image_node(node):
    """Return dict that describes the image defined in the node

    :param node: ElementTree node
    :return: dictionary that describes the image
    """
    d = dict()
    d['source'] = node.get('source')
    d['trans'] = node.get('trans', None)
    d['width'] = int(node.get('width'))
    d['height'] = int(node.get('height'))
    return d


def decode_gid(raw_gid):
    """Decode a GID from data

    as of 0.7.0 it determines if the tile should be flipped when rendered
    as of 0.8.0 bit 30 determines if GID is rotated

    :param raw_gid: 32-bit number from layer data
    :return: gid, flags
    """
    flags = TileFlags(
        raw_gid & GID_TRANS_FLIPX == GID_TRANS_FLIPX,
        raw_gid & GID_TRANS_FLIPY == GID_TRANS_FLIPY,
        raw_gid & GID_TRANS_ROT == GID_TRANS_ROT)
    gid = raw_gid & ~(GID_TRANS_FLIPX | GID_TRANS_FLIPY | GID_TRANS_ROT)
    return gid, flags


def parse_properties(node):
    """Parse a node and return a dict that represents a tiled "property"

    the "properties" from tiled's tmx have an annoying quality that "name"
    and "value" is included. here we mangle it to get that junk out.
    """
    d = dict()
    for child in node.iterfind('properties'):
        for subnode in child.iterfind('property'):
            d[subnode.get('name')] = subnode.get('value')
    return d


def cast_and_set_attributes_from_node_items(object_, items):
    for key, value in items:
        casted_value = types[key](value)
        setattr(object_, key, casted_value)


def contains_invalid_property_name(object_, items):
    for k, v in items:
        if hasattr(object_, k):
            msg = '{0} "{1}" has a property called "{2}"'
            logger.error(msg.format(object_.__class__.__name__, object_, k,
                                    object_.__class__.__name__))
            return True

        if object_.reserved is not None:
            if k in object_.reserved:
                return True

    return False


def set_properties(object_, node):
    """
    read the xml attributes and tiled "properties" from a xml node and fill
    in the values into the object's dictionary.  Names will be checked to
    make sure that they do not conflict with reserved names.
    """
    cast_and_set_attributes_from_node_items(object_, node.items())
    properties = parse_properties(node)
    if contains_invalid_property_name(object_, properties.items()):
        msg = "This name(s) {1} is reserved for {0} objects and cannot be used."
        logger.error(msg.format(object_.__class__.__name__, properties.items()))
        logger.error("Please change the name(s) in Tiled and try again.")
        raise ValueError

    object_.properties = properties


def parse_map(node, tiled_map=None):
    """Parse a map from ElementTree xml node

    :param node: ElementTree xml node
    :return: TiledMap Instance
    """
    if tiled_map is None:
        tiled_map = TiledMap()
    set_properties(tiled_map, node)
    tiled_map.background_color = node.get('backgroundcolor', None)

    # ***         do not change this load order!         *** #
    # ***    gid mapping errors will occur if changed    *** #

    for subnode in node.findall('layer'):
        layer = parse_layer(subnode, tiled_map)
        tiled_map.layernames[layer.name] = layer
        tiled_map.layers.append(layer)
    for subnode in node.findall('imagelayer'):
        tiled_map.layers.append(parse_imagelayer(subnode, tiled_map))
    for subnode in node.findall('objectgroup'):
        tiled_map.layers.append(parse_objectgroup(subnode, tiled_map))
    for subnode in node.findall('tileset'):
        tiled_map.tilesets.append(parse_tileset(subnode, tiled_map))

    # for o in [o for o in tiled_map.objects if o.gid]:
    #     # gid/tile properties are defined in the tileset, but also apply
    #     # to objects, sometimes.  here we get the gid/tile props and update them
    #     p = tiled_map.get_tile_properties_by_gid(o.gid)
    #     if p:
    #         o.properties.update(p)
    #
    #     # tiled stores the origin of GID objects by the lower right corner
    #     # this is different for all other types, so is is flipped here
    #     try:
    #         tileset = tiled_map.get_tileset_from_gid(o.gid)
    #     except ValueError:
    #         msg = 'attempted to lookup invalid gid %s in object %s'
    #         logger.error(msg, o.gid, o)
    #     else:
    #         if tiled_map.invert_y:
    #             o.y -= tileset.tileheight
    #         o.height = tileset.tileheight
    #         o.width = tileset.tilewidth

    tiled_map.reload_images()
    return tiled_map


def parse_objectgroup(node, parent):
    """Parse an Object Group from ElementTree xml node

    :param node: ElementTree xml node
    :return: TiledObjectGroup Instance
    """
    og = TiledObjectGroup()
    set_properties(og, node)
    for child in node.iterfind('object'):
        object_ = parse_object(child)
        og.append(object_)

    return og


def parse_tileset(node, parent):
    """Parse a Tileset from ElementTree xml node

    A bit of mangling is done to allow tilesets with external TSX files

    :param node: ElementTree xml node
    :return: TiledTileset Instance
    """
    tileset = TiledTileset()

    # if true, then node references an external tileset
    source = node.get('source', None)
    if source:
        if source[-4:].lower() == ".tsx":

            # external tilesets don't save the gid, store it for later
            tileset.firstgid = int(node.get('firstgid'))

            # we need to mangle the path b/c tiled stores relative paths
            dirname = os.path.dirname(parent.filename)
            path = os.path.abspath(os.path.join(dirname, source))
            try:
                node = ElementTree.parse(path).getroot()
            except IOError:
                msg = "Cannot load external tileset: {0}"
                logger.error(msg.format(path))
                raise Exception

        else:
            msg = "Found external tileset, but cannot handle type: {0}"
            logger.error(msg.format(tileset.source))
            raise Exception

    set_properties(tileset, node)

    for child in node.getiterator('tile'):
        p = parse_properties(child)
        tiled_gid = int(child.get("id"))

        # handle tiles that have their own image
        image_node = child.get('image', None)
        if image_node:
            p.update(handle_image_node(image_node))
        else:
            p['width'] = tileset.tilewidth
            p['height'] = tileset.tileheight

        tileset.tile_properties[tiled_gid] = p

    # handle the optional 'tileoffset' node
    offset = node.find('tileoffset')
    if offset:
        tileset.offset = (offset.get('x', 0), offset.get('y', 0))
    else:
        tileset.offset = (0, 0)

    image_node = node.find('image')
    if image_node is not None:
        for k, v in handle_image_node(image_node).items():
            setattr(tileset, k, v)

    return tileset


def parse_layer(node, parent):
    """Parse a Tile Layer from ElementTree xml node

    :param node: ElementTree xml node
    :return: TiledTileLayer Instance
    """
    layer = TiledTileLayer()
    set_properties(layer, node)
    data = None
    next_gid = None
    data_node = node.find('data')

    encoding = data_node.get('encoding', None)
    if encoding == 'base64':
        from base64 import b64decode
        data = b64decode(data_node.text.strip())

    elif encoding == 'csv':
        next_gid = map(int, "".join(
            line.strip() for line in data_node.text.strip()
        ).split(","))

    elif encoding:
        msg = 'encoding type: {0} is not supported.'
        logger.error(msg.format(encoding))
        raise Exception

    compression = data_node.get('compression', None)
    if compression == 'gzip':
        import gzip
        with gzip.GzipFile(fileobj=six.BytesIO(data)) as fh:
            data = fh.read()

    elif compression == 'zlib':
        import zlib
        data = zlib.decompress(data)

    elif compression:
        msg = 'compression type: {0} is not supported.'
        logger.error(msg.format(compression))
        raise Exception

    # if data is None, then it was not decoded or decompressed, so
    # we assume here that it is going to be a bunch of tile elements
    # TODO: this will/should raise an exception if there are no tiles
    if encoding == next_gid is None:
        def get_children(parent):
            for child in parent.iterfind('tile'):
                yield int(child.get('gid'))

        next_gid = get_children(data_node)

    elif data:
        if type(data) == bytes:
            fmt = struct.Struct('<L')
            iterator = (data[i:i + 4] for i in range(0, len(data), 4))
            next_gid = (fmt.unpack(i)[0] for i in iterator)
        else:
            msg = 'layer data not in expected format ({0})'
            logger.error(msg.format(type(data)))
            raise Exception

    init = lambda: [0] * layer.width

    # H (16-bit) may be a limitation for very detailed maps
    layer.data = tuple(array.array('H', init()) for i in range(layer.height))
    for (y, x) in product(range(layer.height), range(layer.width)):
        layer.data[y][x] = parent._register_gid(*decode_gid(next(next_gid)))

    return layer


def parse_object(node):
    """Parse an Object from ElementTree xml node

    :param node: ElementTree xml node
    :return: TiledObject Instance
    """

    def read_points(text):
        """parse a text string of float tuples and return [(x,...),...]
        """
        return tuple(tuple(map(float, i.split(','))) for i in text.split())

    object_ = TiledObject()
    set_properties(object_, node)

    points = None
    polygon = node.find('polygon')
    if polygon is not None:
        points = read_points(polygon.get('points'))
        object_.closed = True

    polyline = node.find('polyline')
    if polyline is not None:
        points = read_points(polyline.get('points'))
        object_.closed = False

    if points:
        x1 = x2 = y1 = y2 = 0
        for x, y in points:
            if x < x1: x1 = x
            if x > x2: x2 = x
            if y < y1: y1 = y
            if y > y2: y2 = y
        object_.width = abs(x1) + abs(x2)
        object_.height = abs(y1) + abs(y2)
        object_.points = tuple(
            [(i[0] + object_.x, i[1] + object_.y) for i in points])

    return object_


def parse_imagelayer(node, parent):
    """Parse an Image Layer from ElementTree xml node

    :param node: ElementTree xml node
    :return: TiledImageLayer instance
    """
    image_layer = TiledImageLayer()
    set_properties(image_layer, node)
    image_layer.name = node.get('name', None)
    image_layer.opacity = node.get('opacity', image_layer.opacity)
    image_layer.visible = node.get('visible', image_layer.visible)
    image_node = node.find('image')
    image_layer.source = image_node.get('source')
    image_layer.trans = image_node.get('trans', None)
    return image_layer
